src/node/gov/handlers/acks.h

```diff
@@ -0,0 +1,322 @@
+          // Get memberId from path parameter
+          std::string error;
+          std::string member_id_str;
+          if (!ccf::endpoints::get_path_param(
+                ctx.rpc_ctx->get_request_path_params(),
+                "memberId",
+                member_id_str,
+                error))
+          {
+            detail::set_gov_error(
+              ctx.rpc_ctx,
+              HTTP_STATUS_BAD_REQUEST,
+              ccf::errors::InvalidResourceName,
+              std::move(error));
+            return;
+          }
+
+          // Confirm this matches memberId from signature
+          ccf::MemberId member_id(member_id_str);
+          const auto& cose_ident =
+            ctx.template get_caller<ccf::MemberCOSESign1AuthnIdentity>();
+          if (cose_ident.member_id != member_id)
+          {
+            detail::set_gov_error(
+              ctx.rpc_ctx,
+              HTTP_STATUS_BAD_REQUEST,
+              ccf::errors::InvalidAuthenticationInfo,
+              fmt::format(
+                "Member ID from path parameter ({}) does not match "
+                "member ID from body signature ({}).",
+                member_id,
+                cose_ident.member_id));
+            return;
+          }
+
+          // Check an expected digest for this member is the KV
+          auto acks_handle =
+            ctx.tx.template rw<ccf::MemberAcks>(Tables::MEMBER_ACKS);
+          auto ack = acks_handle->get(member_id);
+          if (!ack.has_value())
+          {
+            detail::set_gov_error(
+              ctx.rpc_ctx,
+              HTTP_STATUS_FORBIDDEN,
+              ccf::errors::AuthorizationFailed,
+              fmt::format("No ACK record exists for member {}.", member_id));
+            return;
+          }
+
+          // Check signed digest matches expected digest in KV
+          const auto expected_digest = ack->state_digest;
+          const auto signed_body = nlohmann::json::parse(cose_ident.content);
+          const auto actual_digest =
+            signed_body["stateDigest"].template get<std::string>();
+          if (expected_digest != actual_digest)
+          {
+            detail::set_gov_error(
+              ctx.rpc_ctx,
+              HTTP_STATUS_BAD_REQUEST,
+              ccf::errors::StateDigestMismatch,
+              fmt::format(
+                "Submitted state digest is not valid.\n"
+                "Expected\n"
+                " {}\n"
+                "Received\n"
+                " {}",
+                expected_digest,
+                actual_digest));
+            return;
+          }
+
+          // Ensure old HTTP signed req is nulled
+          ack->signed_req = std::nullopt;
+
+          // Insert new signature
+          ack->cose_sign1_req = std::vector<uint8_t>(
+            cose_ident.envelope.begin(), cose_ident.envelope.end());
+
+          // Store signed ACK in KV
+          // TODO: Behaviour change: don't automatically update the
+          // state_digest here
```

Comment:
Minor behaviour change, that I think simplifies the API (each endpoint does "one thing"). The old `/ack` endpoint would _also_ do the work of updating and producing a fresh state digest. I think that's unnecessary, and avoiding it lets us return a `204` from this endpoint.

Reply:
Not sure about this to be honest, it turns one user step into two (*), and adds a possible failure case (refresh, fetch from out of date backup, fail to update), which wasn't possible when everything landed on the primary.

If it's what was signed off in the typespec, that's fine of course.

(*) the value of acking as a built-in is not clear to me, I think we would ideally move towards it being something that can be done through the propose-vote-apply flow. For example by allowing (or even requiring) a vote from a newly added member, and transitioning them straight to ACTIVE if they do that. A generic way to do that may be to add an `electors(proposal, ro_kv)` call to the constitution.

Reply:
I don't think it's an extra step. The previous `/ack` would silently update the table containing what it expected a member to ack next, but didn't actually return this to the caller (it returns a `204 No Content`). So the user had to call `/update_state_digest` to retrieve a digest (there was no pure `GET`), causing a _further_ write, to read what they were expected to sign (or somehow get the expected digest from the ledger, but assume that's not the expected path). So with either API, the expected pattern is "call /update to get a 'fresh' (meaningful) digest, then /ack to submit your signature". The new API also has a `GET` for RESTy simplicity, which I guess is more likely to return a stale result and cause an error, but also allows some class of "restarting", if anyone cared.

Fully agree with the goal of removing built-in ACKing, but I think that's out-of-scope here. I think we could implement it with existing calls, as a slight modification of the sample constitution - every member has a proposal they can unilaterally pass to mark themselves as active, only members who have done this are counted in `resolve()`. Can include signing of some KV-stored freshness nonce if desired. Then the service loses any concept of member status/activeness, and instead a constitution which cares can choose to only count votes and proposals from active members.

Reply:
> The previous /ack would silently update the table containing what it expected a member to ack next, but didn't actually return this to the caller (it returns a 204 No Content). 

Ok, I'd missed that and misunderstood what was being split, I agree that shouldn't be bundled.

> Fully agree with the goal of removing built-in ACKing, but I think that's out-of-scope here.

Agreed, I wasn't suggesting it was in scope.

The KV-store freshness happens for free now that proposal ids are a combined digest of the tree root and the proposal, so I think a discrete digest is unnecessary. I think it's still worth exploring how to do consortium changes atomically, rather than keep the part of, but inactive state, which as far as I can tell is only ever used by accident. This is especially useful when rotating a member key without letting the quorum drop.